src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java

```diff
@@ -0,0 +1,344 @@
+ * </p>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name="PreferMethodReference"/&gt;
+ * </pre>
+ * <p>Example of violations:</p>
+ * <pre>
+ * Runnable r = () -&gt; System.out.println(); // violation
+ * BiConsumer&lt;Long, Long&gt; c = (a, b) -&gt; Long.sum(a,b); // violation
+ * </pre>
+ * <p>
+ * Note: This check is not type-aware, so some cases cannot be detected, e.g.
+ * </p>
+ * <pre>
+ * Function&lt;String, Object&gt; f1 =
+ *     a -&gt; (Object) "string".concat(a); // no violation because of cast
+ * </pre>
+ *
+ * @since 8.23
+ */
+@StatelessCheck
+public class PreferMethodReferenceCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_METHOD_REF = "method.reference";
+
+    /**
+     * List of tokens that is used to check expression.
+     *
+     * @see #shouldCheck(DetailAST)
+     */
+    private static final List<Integer> TOKEN_TYPES = Arrays.asList(
+        TokenTypes.EMPTY_STAT,
+        TokenTypes.LITERAL_DO,
+        TokenTypes.LITERAL_FOR,
+        TokenTypes.LITERAL_IF,
+        TokenTypes.LITERAL_RETURN,
+        TokenTypes.LITERAL_SWITCH,
+        TokenTypes.LITERAL_THROW,
+        TokenTypes.LITERAL_TRY,
+        TokenTypes.LITERAL_WHILE
+    );
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {TokenTypes.LAMBDA};
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final List<String> args = getLambdaArgs(ast);
+        final DetailAST expr = ast.getLastChild();
+        final boolean hasViolation = checkExpr(expr, args);
+        if (hasViolation) {
+            log(ast, MSG_METHOD_REF);
+        }
+    }
+
+    /**
+     * Checks if this expression can be replaced with method reference.
+     *
+     * @param expr token to check
+     * @param args lambda arguments
+     * @return true if can be replaced, false otherwise
+     */
+    private static boolean checkExpr(DetailAST expr, List<String> args) {
+        final DetailAST child = expr.getFirstChild();
+        final boolean result;
+        switch (child.getType()) {
+            case TokenTypes.METHOD_CALL:
+                result = checkMethodCall(child, args);
+                break;
+            case TokenTypes.LITERAL_NEW:
+                result = checkObjectCreation(child, args);
+                break;
+            case TokenTypes.SLIST:
+            case TokenTypes.LITERAL_RETURN:
+                result = child.getFirstChild().getType() != TokenTypes.SEMI
+                    && checkExpr(child.getFirstChild(), args);
+                break;
+            default:
+                result = shouldCheck(expr)
```

Comment:
please order method declaration by our future Check - https://github.com/sevntu-checkstyle/methods-distance
to minimize distance between methods (for now it is web service).

Reply:
@rnveach  and @pbludov , Looks like we forgot about our experimental attempt to control even order of methods in class. While we are in progress of searching good formula for order, we try to "enforce" such "Check" over new Modules. Please request this during review of new Modules.

Reply:
I rearranged them, please take a look